aws_lambda_powertools/metrics/metrics.py

```diff
@@ -90,11 +87,41 @@ def __init__(self, service: str = None, namespace: str = None):
             service=self.service,
         )
 
+    def set_default_dimensions(self, **dimensions):
+        """Persist dimensions across Lambda invocations
+
+        Parameters
+        ----------
+        dimensions : Dict[str, Any], optional
+            metric dimensions as key=value
+
+        Example
+        -------
+        **Sets some default dimensions that will always be present across metrics and invocations**
+
+            from aws_lambda_powertools import Metrics
+
+            metrics = Metrics(namespace="ServerlessAirline", service="payment")
+            metrics.set_default_dimensions(environment="demo", another="one")
+
+            @metrics.log_metrics()
+            def lambda_handler():
+                   return True
+        """
+        self.default_dimensions.update(**dimensions)
+        for name, value in dimensions.items():
+            self.add_dimension(name, value)
+
+    def clear_default_dimensions(self):
+        self.default_dimensions.clear()
+
     def clear_metrics(self):
         logger.debug("Clearing out existing metric set from memory")
         self.metric_set.clear()
         self.dimension_set.clear()
         self.metadata_set.clear()
+        # re-add default dimensions
+        self.dimension_set.update(**self._default_dimensions)
```

Comment:
It looks like _default_dimensions is not updated (in 79 you are updating self.default_dimensions, not the other way around), so it looks like it may be empty at this point? - Also, in L111-L113 you are calling the self.add_dimensions() inside the set_default_dimensions() but once dimension_set is cleared if clear_metrics() is called you are not calling add_dimensions again, so it would look like this will be empty on a second invocation. Can't test at the moment, but that what it looks like from a quick overview of the code.

Reply:
Yeah, it's a bug as dimensions aren't added in the right format - Thanks for catching it!

Everything else isn't empty because we're "abusing" class instance vars similarly to what the Borg pattern does.

L72 `_default_dimensions: Dict[str, Any] = {}`, here we initialize it as an empty dict and as class instance so if we were to initialize a new instance next time it'll keep the state.